### ISSUE TITLE
finalizer deletion bug in k8s needs to be worked around in ansible operator

### ISSUE THREAD

**Issue by jmazzitelli** [operator-framework/operator-sdk#1493]
(for a complete set of replication procedures including code and scripts to run to see the problem, see my comment down below: https://github.com/operator-framework/operator-sdk/issues/1493#issuecomment-497485434)

This k8s bug (https://github.com/kubernetes/kubernetes/issues/73098) has hit both the [Maistra](https://issues.jboss.org/browse/MAISTRA-378) and [Kiali](https://issues.jboss.org/browse/KIALI-2770) operators (and, actually, this probably affects all operators).

The bug is simply:

1. Create namespace N, and in that namespace install an operator and the CR that triggers the installation of the resources it manages.
2. Delete namespace N where the operator/CR lives
3. The deletion hangs.

 Maistra worked around it in their Go operator via this loop that continues to retry [using this code](https://github.com/Maistra/istio-operator/blob/maistra-0.11/pkg/controller/servicemesh/controlplane/controller.go#L143-L158).

The ansible operator framework needs to incorporate something similar - otherwise, all ansible-based operators will hit this bug without an easy workaround. It looks like a solution like how Maistra did it (with a retry loop) should be done somewhere in or around here: https://github.com/operator-framework/operator-sdk/blob/v0.8.x/pkg/ansible/controller/reconcile.go#L97-L107

Note that the underlying k8s bug that is causing all the problems is fixed in k8s 1.14+ but that doesn't help OpenShift 4.0 and current 4.1 users. It also affects OpenShift 3.11 users.

**Comment by camilamacedo86** [comment]
Hi @jmazzitelli if you check the https://github.com/kubernetes/kubernetes/issues/73098 shows that the issue there was caused because of the way that it was used. Following an explanation which I hope that helps you. 

If you add a finalizer means that it is a flag and the CR will not be deleted until you remove this flag since the idea of using it is just for you have a way to perform actions before the CR be deleted from the cluster. For example, you can impl to allow deleted it just after call a REST endpoint and do something. 

After doing what s required to get done you need to update the CR and remove the finalizer from there then it will be deleted. as follows. 

```
// TODO(user): Add the cleanup steps that the operator needs to do before the CR can be deleted
 		// Update finalizer to allow delete CR
 		memcached.SetFinalizers(nil)
 
 		// Update CR
 		err := r.client.Update(context.TODO(), memcached)
 		if err != nil {
 			return reconcile.Result{}, err
 		}
 		return reconcile.Result{}, nil
```

I'd like to recommend you see this doc: https://github.com/operator-framework/operator-sdk/blob/2e1c2bc3cdbe8c08a7b29e721e5ae635803adfbc/doc/user-guide.md#handle-cleanup-on-deletion

Also, you can check in the project that I am working on a place where I am using the finalizer. See [here] (https://github.com/aerogear/mobile-security-service-operator/blob/master/pkg/controller/mobilesecurityserviceapp/controller.go#L169)

**Comment by jmazzitelli** [comment]
@camilamacedo86 this issue is specifically for the ansible operator framework. The link you provided is to code implemented directly in Go; the doc you recommended is specifically for operators implemented in Go. So those are not really applicable to the issue I'm documenting here (they will be applicable to the folks that need to fix this in the ansible operator framework/sdk - but for the purposes of this github issue - I'm talking about how Ansible operators are going to hit this problem).

Here is the Ansible Operator docs - it explains the recommended finalizer usage in ansible operators when needing to delete resources:

https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/finalizers.md

I use finalizers as that doc recommends (causes the ansible operator to run clean up code).

You can read about what problems the operator-sdk had for Go-implemented operators by reading what happened with the Maistra operator and how they went about working around this k8s finalizer problem - see: https://issues.jboss.org/browse/MAISTRA-378

We need the Ansible Operator framework to implement something similar because ansible-based operators are going to hit this problem (well, they ARE hitting this problem) as we've already seen on OpenShift 4 with the Kiali operator (an Ansible-implementation).

**Comment by camilamacedo86** [comment]
Hi @jmazzitelli,

OK. But no matter the type of operator project the finalizers at the end will work in the same way. I mean it will be a metadata tag added in the CR in your cluster which will not allow it to be deleted until it be removed.

If you look at the doc provided by you, shows that the action which need be performed which when finishing with success will allow deleting it is here `/opt/ansible/destroy.yml`. Are you doing it? Did you check if the steps added in this part are working with success in your scenario? 

PS.: The code impl [here](https://github.com/operator-framework/operator-sdk/blob/v0.8.x/pkg/ansible/controller/reconcile.go#L98) sent for you is justing adding the finalizer metadata according to the configs in the same way that it has done in the example shared above for the golang operator. I mean it will check if the GetDeletionTimestamp() <> "" which means that the CR was marked to be deleted and if has a finalizer to added on the CR will perform it in the first steps of the reconcile. 

In this way, I'd recommend to you:

- Re-check your implementation. Ensure that the action called when the app CR will be deleted has been done with success. (I think that here is your problem)
- If you are not able to figure out what is happing, I'd recommend updating the first comment to make clear your problem. I'd suggested using its template in order to provide all required info
- If you are sure that it is an issue which will be solved with the upper version of kubernetes 1.14+ the good news is that controller-runtime the latest beta version is with kubernetes 1.14+ which shows that it would be upgraded in a soon future as well.

**Comment by jmazzitelli** [comment]
The implementation works on OpenShift 3.11. We only see this on OpenShift 4. So I'll assume my ansible playbooks are correctly implemented.

But, yes, I'm fairly confident that my playbooks are doing it how the docs say - in fact, I even make sure my ansible tasks NEVER end with an error (I have "ignore_errors: yes" in every one of my tasks) to ensure my ansible remove playbook ALWAYS finish with a success. Because, in fact, if the ansible playbook fails in any way, the ansible operator framework will not remove the finalizer tag and the hang will also happen (but this is to be expected). This is not what is happening AFAIK when on OpenShift 4.

**Comment by jmazzitelli** [comment]
This is how Maistra worked around the problem - it is this workaround that most likely needs to be implmemented in the operator-sdk Go code that underlies the ansible operator framework code:

https://github.com/Maistra/istio-operator/blob/maistra-0.11/pkg/controller/servicemesh/controlplane/controller.go#L143-L158

cc @rcernich - Rob knows more about the underlying problem and solution.

**Comment by rcernich** [comment]
@camilamacedo86, this is an issue if the user deletes the namespace containing the custom resource.  In that case, k8s updates the deletion timestamp every time it runs though its event loop.  If this happens before the operators processing is complete, you'll get a conflict error when you go to remove the finalizer, which effectively blocks the namespace and resource from being deleted.  What's worse is that after an initial grace period, the deletion timestamp gets updated every second, which makes it impossible for an end user to remove the finalizer without using the `patch` mechanism.

If you haven't seen this issue before, then most likely your finalization is completing before the grace period.

I can also say that this problem occurs on 3.11 too, and that you won't see the problem if your finalization process completes before the grace period.

**Comment by camilamacedo86** [comment]
Hi @jmazzitelli,

I'd recommend you update the first comment and make it short, and clear as possible by following its template. I think it will help others understand better your issue. Note that the info over versions used and etc are quite important. 

 ## Bug Report

**What did you do?**
I did a similar impl as described in [ansible/dev/finalizers.md](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/dev/finalizers.md)

**What did you expect to see?**
The implementation works on OpenShift 3.11. We only see this on OpenShift 4. So I'll assume my ansible playbooks are correctly implemented and I expected seen it working in the same way. 

**What did you see instead? Under which circumstances?**
When I deleted the namespace where the CR is applied the deletion hangs. Following the steps performed. 

> 1. Create namespace N, and in that namespace install an operator and the CR that triggers the installation of the resources it manages.
> 2. Delete namespace N where the operator/CR lives
> 

**Environment**
* operator-sdk version: `<!--- Insert operator-sdk release or Git SHA here. -->`

* go version:`<!--- Insert the output of `go version` here -->`
* Kubernetes version information: `<!--- Insert the output of `kubectl version` here -->`
* Kubernetes cluster kind: OpenShift 4

* Are you writing your operator in ansible, helm, or go?
ansible

**Possible Solution**
Upgrade the operator project to use the k8s 1.14+ since shows that it caused by kubernetes/kubernetes#73098

**Additional context**

The ansible operator framework needs to incorporate something similar - otherwise, all ansible-based operators will hit this bug without an easy workaround. It looks like a solution like how Maistra did it (with a retry loop) should be done somewhere in or around here: https://github.com/operator-framework/operator-sdk/blob/v0.8.x/pkg/ansible/controller/reconcile.go#L97-L107

Note that the underlying k8s bug that is causing all the problems is fixed in k8s 1.14+ but that doesn't help OpenShift 4.0 and current 4.1 users. **This does not affect OpenShift 3.11 users.**

**Comment by rcernich** [comment]
if you want to do a test, i suspect just putting a sleep call in your finalization processing will successfully reproduce the issue.

**Comment by jmazzitelli** [comment]
@rcernich just informed me he saw this on OpenShift 3.11 as well.. I haven't seen it, but if its a timing issue, that could be why.

**Comment by jmazzitelli** [comment]
I'm going to post a very simple Ansible operator that replicates this pretty easily. Stay tuned.

**Comment by jmazzitelli** [comment]
I am attaching a zip with a simple ansible operator to replicate this. It is extraordinarily small. Here it is:

[optest.zip](https://github.com/operator-framework/operator-sdk/files/3238665/optest.zip)

The `watches.yaml` looks like:

```
- version: v1alpha1
  group: optest.example.com
  kind: Optest
  role: /opt/ansible/roles/optest
  reconcilePeriod: 0
  finalizer:
    name: finalizer.optest
    role: /opt/ansible/roles/optest-remove
```

The only task in the optest role is this (so this is triggered when the CR is created):

```
- debug:
    msg: STARTING OPTEST
```

When the CR is removed, the optest-remove role is triggered (notice how small it is, and that all errors that might occur are ignored. All this has is a debug and pause task - so should never fail anyway. So this will never fail but will take 60 seconds to complete):

```
- debug:
    msg: "GOING TO SLEEP"
  ignore_errors: yes
- pause:
    seconds: 60
  ignore_errors: yes
- debug:
    msg: "AWAKE NOW"
  ignore_errors: yes
```

In the zip are some scripts to make running this test easy. Here is how to see the problem, run these commands in this order (this assumes your cwd is in the top directory - the "optest" directory - the first step does this):

1. Unzip and go to the optest directory:
```
unzip optest.zip
cd optest
```

2. Build the test operator image and store in your docker repo labeled as "jmazzitelli/optest:latest":
```
./build-operator.sh
```

3. Create the operator's namespace (it is called `optestns`):
```
./create-namespace.sh
```

4. Create the operator resources - this starts the operator:
```
./create-operator.sh
```

5. Wait for the operator to start - just keep running `oc get deployment optest -n optestns` until the pod is available and started.

6. Now install a CR to trigger the operator's ansible optest role:
```
./create-cr.sh
```

7. OK, now the problem. Wait a little while for the operator to complete processing the new CR you created in the previous step. Now delete the namespace. Watch how it hangs - the Ansible operator's remove role is triggered, and it takes 60 seconds to complete, but the finalizer failed to be removed by the Ansible operator framework code (remember, this is not the job of the Ansible operator itself - the underlying operator framework does this for us), so the namespace can never go away:
```
./delete-namespace.sh
```

8. At this point, you are screwed until you manually patch the CR to remove the finalizer tag. Notice the CR still exists, still has the finalizer tag AND notice the delete timestamp keeps getting updated - run this a bunch of times to see:

```
oc get optest example-optest -n optestns -o yaml
```

9. In order to delete this CR and namespace, you must manually patch the CR to remove the finalizer - this script will do that:
```
./patch-cr.yaml
```

Notice how quickly your "./delete-namespace.sh" wakes up from its hang and finishes the deletion of the namespace?? The namespace is deleted now. You can remove the CRD by running "./delete-operator.sh" - that will also remove the CRD.

The .sh scripts are simple - many are one-liners. I just made them .sh for it to be easily consumable and repeatable. There is nothing in the sh scripts but simple "oc" commands to create and delete resources.

You can confirm the remove role actually works if you remove the CR first before you try to delete the namespace. So start from the beginning, and after you run `./create-cr.sh` you can then run `./delete-cr.sh`. Notice the Ansible operator logs shows the debug messages found in the optest-remove role task. Now you can `./delete-namespace.sh` and the deletion works fine - the namespace and the rest of the operator resources are removed successfully without hanging.

**Summary**

In summary, as you can see by this trivial example, there is nothing happening in the removal role that would cause the finalizer to fail to be removed. As @rcernich points out, this appears to be a timing issue due to that k8s bug and requires the Ansible framework itself to workaround it in the same manner as I posted earlier (pointing to how the Maistra Go code works around it). Hence what this github issue is asking for - a workaround to this k8s bug.

Note that I replicated this on OpenShift 3.11 and we've see it also on OpenShift 4 - so it's a problem in both.

**Comment by jmazzitelli** [comment]
@camilamacedo86  @rcernich see my complete replication procedures documented in the previous comment - has an attachment with the trivial operator code that shows the bug

**Comment by rcernich** [comment]
@jmazzitelli, thanks for working up a concise reproducer.  @camilamacedo86, hopefully the reproducer helps you better understand the issue.

**Comment by camilamacedo86** [comment]
Hi @jmazzitelli, 

## Root Cause

**The issue faced has no relation with the OCP version and/or ks8 lib at all.** 

The root cause of it is because the CR has been created without the labels which are allowing the operator to be deleted before its finalizer be executed and the deletion of the CR. 

<img width="909" alt="Screenshot 2019-05-31 at 01 52 10" src="https://user-images.githubusercontent.com/7708031/58674245-cbb38500-8346-11e9-92f3-2fca6611b20f.png">

Note that the finalizer will work if you delete the C by running `oc delete -n optestns -f deploy/crds/optest_v1alpha1_optest_cr.yaml`

### Why the issue is faced?

It can be some issue in the config/impl of your [optest.zip](https://github.com/operator-framework/operator-sdk/files/3238665/optest.zip) project and or the operator-framework which is not adding the respective labels in the CR.

### So, how we can check it? How to isolate the issue? 

We need to use the master branch to create an ansible Memcached example project by following the [doc](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/user-guide.md#create-a-new-project) and check if the labels of its CR will no be added too. 

## Tests performed. 

Following the test executed to ensure that the finalizer is working. 

1. The watch was changed as follows. 

```yaml
---
- version: v1alpha1
  group: optest.example.com
  kind: Optest
  role: /opt/ansible/roles/optest
  finalizer:
    name: optest.example.com  # the finalizer name to follow the convecton <finalizer>-<group> 
    role: /opt/ansible/roles/optest-remove
```
NOTE: I removed ` reconcilePeriod: 0`.

2 The CR was removed by running `oc delete -n optestns -f deploy/crds/optest_v1alpha1_optest_cr.yaml`

NOTES: 
* You can check the CR by going to "resources>>others resources>> select it"
* You can be monitoring the operator actions my checking its logs of this pod

**Comment by camilamacedo86** [comment]
Hi @jmazzitelli, 

## Root Cause

The issue faced shows has no relation with the OCP version and/or ks8 lib at all. It is occurring because the operator has been deleted before the CR. **Note that the finalizer will work if you delete the CR by running `oc delete -n optestns -f deploy/crds/optest_v1alpha1_optest_cr.yaml`**

### Why the issue is faced?

It can be some issue in your [optest.zip](https://github.com/operator-framework/operator-sdk/files/3238665/optest.zip) project or a bug in the operator-framework. Required further analysis and tests. 

### What are the next steps? 

We need to use the master branch to create an ansible Memcached example project by following the [doc](https://github.com/operator-framework/operator-sdk/blob/master/doc/ansible/user-guide.md#create-a-new-project) and check if the same behaviour will be faced or not with. 

## Tests performed. 

**Following the test executed to ensure that the finalizer is working.** 

1. The watch was changed as follows. 

```yaml
---
- version: v1alpha1
  group: optest.example.com
  kind: Optest
  role: /opt/ansible/roles/optest
  finalizer:
    name: optest.example.com  #I changed the finalizer name to follow the convention <finalizer>-<group> 
    role: /opt/ansible/roles/optest-remove
```
NOTE: I removed ` reconcilePeriod: 0`.

2 The CR was removed by running `oc delete -n optestns -f deploy/crds/optest_v1alpha1_optest_cr.yaml`

NOTES: 
* You can check the CR by going to "resources>>others resources>> select it"
* You can be monitoring the operator actions my checking its logs of this pod


<img width="1593" alt="Screenshot 2019-05-31 at 02 24 39" src="https://user-images.githubusercontent.com/7708031/58675244-4aaabc80-834b-11e9-8ab6-1fc6a5111254.png">
<img width="863" alt="Screenshot 2019-05-31 at 02 24 32" src="https://user-images.githubusercontent.com/7708031/58675245-4aaabc80-834b-11e9-85e8-2d4a3192afe2.png">

**Comment by jmazzitelli** [comment]
I'm not following the previous comment completely (are you saying you agree there is a problem here, or do you disagree that there is a problem?), but I will confirm that, yes, if you delete the CR first, all works well (as I mention in my earlier comment, _"You can confirm the remove role actually works if you remove the CR first before you try to delete the namespace"_).

But as an internal RedHat QE team has reported to us (and failed their test criteria because of it), if you attempt to delete the operator namespace (with the intent of cleaning up everything in it) without first deleting the CR, then we hit the problem. You can see QE's test-rejection JIRAs [here](https://issues.jboss.org/browse/MAISTRA-378) and [here](https://issues.jboss.org/browse/KIALI-2770). 

Note that once the deletion request to the namespace has been made, you cannot simply ask to delete the CR and expect the namespace deletion to finish. At this point, deleting the CR also hangs. You must patch the CR first (manually removing the finalizer) via:

```
oc patch optest example-optest -n optestns -p '{"metadata":{"finalizers": []}}' --type=merge
```

Once you do this patch, the deletion of the namespace immediately finishes - the namespace is finally removed as well as the last remnants that remained (such as the CR itself).

**Comment by camilamacedo86** [comment]
HI @jmazzitelli,

In order to try to help you, I opened a new Issue with a more clean and assertive description of the issue faced by you.  See https://github.com/operator-framework/operator-sdk/issues/1503

Wdyt over we close this one in favour of https://github.com/operator-framework/operator-sdk/issues/1503 which has a more clean and assertive description over your problem now that we could create a POC to reproduce it and check what is happening?

**Comment by jmazzitelli** [comment]
Closing in favor of the other issue.
